advanced_topics.qmd

```diff
@@ -5,13 +5,114 @@
+
+## Acero bindings
+
+In [Chapter @sec-datamanipulation], we mentioned that the arrow R package transforms dplyr syntax into Arrow Expressions which can be understood by Acero, and here we're going to dive into that a little deeper.
+
+Acero, arrow's query engine, has a large number of functions implemented for performing calculations on data.
+The complete list can be found in the [project documentation](https://arrow.apache.org/docs/cpp/compute.html#available-functions).
+
+Some of these functions have the same names are the R equivalent, for example, both Acero and R have a `min()` function which can be used to find the lowest value in an vector or array of values.
+
+Other functions have slightly different names; whereas R has `sd()` for calculating the standard deviation of a vector, Acero has `stddev()`.
+
+| Feature                        | R (`sd`)                                                | Acero (`stddev`)                                                                        |
+|-------------------|----------------------|--------------------------------|
+| **Function Name**              | `sd`                                                    | `stddev`                                                                                |
+| **Function Options**           | `na.rm = FALSE`                                         | `ddof = 0`, `skip_nulls = FALSE`, `min_count = 1`                                       |
+| **Purpose**                    | Computes standard deviation                             | Computes standard deviation                                                             |
+| **Handling of Missing Values** | `na.rm`: Boolean, removes `NA` values (default `FALSE`) | `skip_nulls`: Boolean, removes `NULL` values (default `FALSE`)                          |
+| **Degrees of Freedom**         | Not available                                           | `ddof`: Integer, denominator degrees of freedom (default `0`)                           |
+| **Minimum Count**              | Not available                                           | `min_count`: Integer, returns `NULL` if fewer non-null values are present (default `1`) |
+
+There are also some differences in the options for these functions; while R's `sd` function has arguments `x`, the value to be passed in and `na.rm`, whether to remove `NA` values or not, Acero's stddev also has `ddof` - denominator degrees of freedom, `skip_nulls`, the equivalent of `na_rm`, and `min_count` - if there are fewer values than this, return `NULL`.
+
+This also highlights the slight different in terminology here; whereas R has separate concepts of `NA` and `NULL`, Arrow considers them to be the same things.
+
+You don't usually have to worry about these differences yourself though, as the code in the arrow R package does a direct translation from one to the other.
```

Comment:
This would be a good place to do a query that has `sd(x)` in it, and then print the query, which prints the C++ option naming.

Reply:
Is this a bug or an unimplemented feature? Having difficulty constructing a query which can do this without me having to also introduce `show_query()`. Here's what I tried:

```r
open_dataset("data/person/") |> 
  summarise(sd_age = sd(AGEP))
```

gets me

```
FileSystemDataset (query)
sd_age: double

See $.data for the source Arrow object
```

I was wondering if it was just aggregations, but apparently not:

```r
open_dataset("data/person/") |> 
  transmute(sd_age = sd(AGEP))
```

gets 

```
FileSystemDataset (query)
sd_age: double

See $.data for the source Arrow object
```

I also tried doing some weird thing with filters which doesn't reeeeally make sense, but that doesn't work either:

```r
open_dataset("data/person/") |> 
  filter(AGEP > sd(AGEP))
```

```
Show in New Window
Error in AGEP > sd(AGEP) : 
  Expression not supported in filter() in Arrow
→ Call collect() first to pull data into R.
```

Reply:
Hmm, I think `sd` is special in some way here as if I swap it for `as.integer()` we get the output:

```
> open_dataset("data/person/") |> 
+     transmute(as.integer(AGEP))
FileSystemDataset (query)
as.integer(AGEP): int32 (cast(AGEP, {to_type=int32, allow_int_overflow=false, allow_time_truncate=false, allow_time_overflow=false, allow_decimal_truncate=false, allow_float_truncate=true, allow_invalid_utf8=false}))
```

Reply:
`sd()` isn't special. Aggregations are just handled differently. The aggregation step happens and then there is a `collapse()`, so the thing you want is inside `$.data`. 

```
> query <- ds |> summarise(sd_age = sd(AGEP))
> query
FileSystemDataset (query)
sd_age: double

See $.data for the source Arrow object
> query$.data
FileSystemDataset (query)
AGEP: int32

* Aggregations:
sd_age: stddev(AGEP)
See $.data for the source Arrow object
```

But for whatever reason (well actually, see `format_aggregation()` in dplyr-summarize.R), aggregations don't print their options in the arrow_dplyr_query print method. But `show_query()` shows them.

```
> show_query(query)
ExecPlan with 4 nodes:
3:SinkNode{}
  2:ScalarAggregateNode{aggregates=[
	stddev(sd_age, {ddof=1, skip_nulls=false, min_count=0}),
]}
    1:ProjectNode{projection=["sd_age": AGEP]}
      0:SourceNode{}
>
```

This is "advanced topics" after all, so maybe `show_query()` isn't a bad thing to introduce. Or, use a scalar function instead of an aggregation as the example here. (Or both! We are short on pages!)

Reply:
Added!